test(flink): add unit test for window join

chloeh13q opened the pull request:
<!--
Thanks for taking the time to contribute to Ibis!

Please ensure that your pull request title matches the conventional commits
specification: https://www.conventionalcommits.org/en/v1.0.0/
-->

## Description of changes

<!--
Write a description of the changes commensurate with the pull request's scope.

Extremely small changes such as fixing typos do not need a description.
-->

1. Added unit test for window join
2. Cleaned up some existing test files
3. Fixed a couple of bugs

A couple of notes:
1. The generated SQL is not an exact match to the example Flink has in their documentation. The one Ibis generates is a little more complicated, because it automatically turns them into subqueries and it's also not able to fuse the select statements automatically.
2. Right now only full outer join works. Other types of join have a different syntax in Flink, e.g. semi join looks like
```sql
SELECT *
FROM (
    SELECT * FROM TABLE(TUMBLE(TABLE LeftTable, DESCRIPTOR(row_time), INTERVAL '5' MINUTES))
) L WHERE L.num IN (
    SELECT num FROM (   
    SELECT * FROM TABLE(TUMBLE(TABLE RightTable, DESCRIPTOR(row_time), INTERVAL '5' MINUTES))
    ) R WHERE L.window_start = R.window_start AND L.window_end = R.window_end);
```
or
```sql
SELECT *
FROM (
    SELECT * FROM TABLE(TUMBLE(TABLE LeftTable, DESCRIPTOR(row_time), INTERVAL '5' MINUTES))
) L WHERE EXISTS (
    SELECT * FROM (
    SELECT * FROM TABLE(TUMBLE(TABLE RightTable, DESCRIPTOR(row_time), INTERVAL '5' MINUTES))
    ) R WHERE L.num = R.num AND L.window_start = R.window_start AND L.window_end = R.window_end);
```
which is not as simple as changing the keyword from `FULL OUTER JOIN` to `LEFT SEMI JOIN`. I need to spend some time digging and understanding how we can support semi/anti window joins.

cpcloud commented on ibis/backends/flink/datatypes.py:
```diff
@@ -88,7 +88,9 @@ def from_ibis(cls, dtype: dt.DataType) -> DataType:
         elif dtype.is_timestamp():
-            return DataTypes.TIMESTAMP(nullable=dtype.nullable)
+            return DataTypes.TIMESTAMP(
+                precision=dtype.scale if dtype.scale else 6, nullable=dtype.nullable
```
```suggestion
                precision=dtype.scale if dtype.scale is not None else 6, nullable=dtype.nullable
```

since `dtype.scale` can be `0` which means seconds precision.

cpcloud commented on ibis/backends/conftest.py:
```diff
@@ -25,6 +25,8 @@
 if TYPE_CHECKING:
     from collections.abc import Iterable
 
+    from ibis.backends.datafusion.tests.conftest import BackendTest
```
I think you probably want a non-backend-specific class here:

```suggestion
    from ibis.backends.tests.base import BackendTest
```

cpcloud commented on ibis/backends/flink/tests/conftest.py:
```diff
@@ -135,3 +180,14 @@ def temp_view(con) -> str:
+        con.drop_table(table, force=True)
+    views_to_drop = list(set(con.list_views()) - set(TEST_TABLES.keys()))
+    for view in views_to_drop:
+        con.drop_view(view, temp=True, force=True)
```
Hm, what is this `temp` argument to `drop_view`? It's a bit odd to have to specify the lifetime of a view when dropping it. Is that a flink idiosyncracy?

chloeh13q replied:
When tables are created from in-memory objects in Flink (e.g. from a pandas dataframe), a temporary view is created. Therefore, when resetting the connection, we also need to explicitly drop the temporary view that was created as well

cpcloud replied:
Right, what I am wondering is why `con.drop_view(view, force=True)` doesn't suffice for that. Why is the additional `temp=True` needed?

chloeh13q replied:
Sorry I misunderstood your question. If you do `con.drop_view(view)` you would get the following error: 
```
org.apache.flink.table.api.ValidationException: Temporary view with identifier '`default_catalog`.`default_database`.`temp`' exists. Drop it first before removing the permanent view.
```

```python
> import pandas as pd
> import ibis
> from pyflink.table import EnvironmentSettings, TableEnvironment
> from pyflink.common import Configuration

> env_settings = EnvironmentSettings.in_streaming_mode()
> table_env = TableEnvironment.create(env_settings)

> table_config = table_env.get_config()
> config = Configuration()
> config.set_string("parallelism.default", "1")
> table_config.add_configuration(config)

> con = ibis.flink.connect(table_env)

> df = pd.DataFrame(
        {
            "row_time": [
                pd.to_datetime("2020-04-15 12:02"),
                pd.to_datetime("2020-04-15 12:06"),
                pd.to_datetime("2020-04-15 12:03"),
            ],
            "num": [1, 2, 3],
            "id": ["L1", "L2", "L3"],
        }
    )
> table = con.create_table("temp", df, temp=True)
> print(f"Tables: {con.list_tables(temp=True)}")
> print(f"Views: {con.list_views()}")
> print(f"Temp views: {con.list_views(temp=True)}")
Tables: ['temp']
Views: ['temp']
Temp views: ['temp']

> con.drop_view("temp")
Py4JJavaError: An error occurred while calling o8.executeSql.
: org.apache.flink.table.api.ValidationException: Temporary view with identifier '`default_catalog`.`default_database`.`temp`' exists. Drop it first before removing the permanent view.
	at org.apache.flink.table.catalog.CatalogManager.dropTableInternal(CatalogManager.java:1231)
	at org.apache.flink.table.catalog.CatalogManager.dropView(CatalogManager.java:1218)
	at org.apache.flink.table.operations.ddl.DropViewOperation.execute(DropViewOperation.java:73)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:1092)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:735)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.flink.api.python.shaded.py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at org.apache.flink.api.python.shaded.py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at org.apache.flink.api.python.shaded.py4j.Gateway.invoke(Gateway.java:282)
	at org.apache.flink.api.python.shaded.py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at org.apache.flink.api.python.shaded.py4j.commands.CallCommand.execute(CallCommand.java:79)
	at org.apache.flink.api.python.shaded.py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.base/java.lang.Thread.run(Thread.java:829)

> con.drop_view("temp", force=True)
Py4JJavaError: An error occurred while calling o8.executeSql.
: org.apache.flink.table.api.ValidationException: Temporary view with identifier '`default_catalog`.`default_database`.`temp`' exists. Drop it first before removing the permanent view.
	at org.apache.flink.table.catalog.CatalogManager.dropTableInternal(CatalogManager.java:1231)
	at org.apache.flink.table.catalog.CatalogManager.dropView(CatalogManager.java:1218)
	at org.apache.flink.table.operations.ddl.DropViewOperation.execute(DropViewOperation.java:73)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:1092)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeSql(TableEnvironmentImpl.java:735)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.flink.api.python.shaded.py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at org.apache.flink.api.python.shaded.py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at org.apache.flink.api.python.shaded.py4j.Gateway.invoke(Gateway.java:282)
	at org.apache.flink.api.python.shaded.py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at org.apache.flink.api.python.shaded.py4j.commands.CallCommand.execute(CallCommand.java:79)
	at org.apache.flink.api.python.shaded.py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.base/java.lang.Thread.run(Thread.java:829)

> con.drop_view("temp", temp=True)

> con.list_views()
[]

> con.list_views(temp=True)
[]
```

ncclementi commented:
@chloeh13q CI seems to be failing, is this one close to done or what's the status?

chloeh13q commented:
@ncclementi Hey Naty, thanks for checking in! I just fixed the CI error and rebased to latest main. It looks like there's now a doc test error but that should be an unrelated error / intermittent failure.

I will respond to @cpcloud 's comment above in the thread, but this PR should be ready for merge otherwise.

ncclementi commented:
I have a fix for the doctest in this PR. https://github.com/ibis-project/ibis/pull/8175
and a fix for the flink CI failure (I think) here https://github.com/ibis-project/ibis/pull/8176.

chloeh13q commented:
@ncclementi Thank you! The fix for the CI test failure LGTM. I also investigated and came up with the same fix and have pushed that in the commit. (p.s. I'm not sure what happened that resulted in this CI failure, it does look like Mehmet had added the flag for flink in his PR 😅)

ncclementi commented:
@chloeh13q Philip just merged all the CI fixes to main, so if you rebase again, you should get a green CI.

chloeh13q commented:
@ncclementi done!

chloeh13q commented:
@cpcloud I believe all the comments have been addressed and resolved. Would you mind taking another look at the PR and approving? Thanks!
